CaesarCipher/CaesarCipher/main.m

```diff
@@ -10,42 +10,81 @@
 
 @interface CaesarCipher : NSObject
 
-- (NSString *)decode:(NSString *)string offset:(int)offset;
 - (NSString *)encode:(NSString *)string offset:(int)offset;
+- (NSString *)decode:(NSString *)string offset:(int)offset;
+
+//Being amateur codebreakers, we want to know if two distinct looking ciphers correspond to the same input message. Write a method called codeBreaker, which accepts two cipher strings as paramaters and returns a boolean value which tells us whether they are actually the same input message encoded using two different offsets.
+
+- (BOOL) codeBreaker:(NSString *) string1 with:(NSString *) string2;
 
 @end
 
+
 @implementation CaesarCipher
 
 - (NSString *)encode:(NSString *)string offset:(int)offset {
     if (offset > 25) {
         NSAssert(offset < 26, @"offset is out of range. 1 - 25");
     }
+    NSString *str = [string lowercaseString];
     unsigned long count = [string length];
     unichar result[count];
     unichar buffer[count];
-    [string getCharacters:buffer range:NSMakeRange(0, count)];
-
+    [str getCharacters:buffer range:NSMakeRange(0, count)];
+    
+    char allchars[] = "abcdefghijklmnopqrstuvwxyz";
+    
     for (int i = 0; i < count; i++) {
         if (buffer[i] == ' ' || ispunct(buffer[i])) {
             result[i] = buffer[i];
             continue;
         }
-
-        int low = islower(buffer[i]) ? 'a' : 'A';
-        result[i] = (buffer[i]%low + offset)%26 + low;
+        
+        char *e = strchr(allchars, buffer[i]);
+        int idx= (int)(e - allchars);
+        int new_idx = (idx + offset) % strlen(allchars);
+        
+        result[i] = allchars[new_idx];
     }
-
+    
     return [NSString stringWithCharacters:result length:count];
 }
 
 - (NSString *)decode:(NSString *)string offset:(int)offset {
     return [self encode:string offset: (26 - offset)];
 }
+
+- (BOOL) codeBreaker:(NSString *) string1 with:(NSString *) string2{
+
+                                                                            //Loop through offset on first encoded input and decode
+    for (int i = 1; i<25; i++) {
+        NSString * decoded = [self decode:string1 offset:i];
+                                                                            //Loop through offset on second encoded input and decode
+        for (int j = 1; j < 25; j++) {
+            NSString * decoded2 = [self decode:string2 offset:j];
+            
+            if ([decoded isEqualToString:decoded2]) {
+                NSLog(@"%@ = '%@', offset: %d", string1, decoded, i);
+                NSLog(@"%@ = '%@', offset: %d", string2, decoded2, j);
+                return YES;
+            }
+            
+        }
+        
```

Comment:
This code is technically correct, however, it can be optimized into a single loop:

```objective-c

    for (int i = 1; i<25; i++) {
            NSString * decoded2 = [self decode:string2 offset:i];
            if ([string1 isEqualToString:decoded2]) {
                return YES;
            }
    }
    
    return NO;
```

The difference is wost case 25 iterations vs. worst case 600+ iterations.